File: core/src/com/nopalsoft/thetruecolor/leaderboard/Person.java
package com.nopalsoft.thetruecolor.leaderboard;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.scenes.scene2d.Group;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Label.LabelStyle;
import com.badlogic.gdx.scenes.scene2d.utils.TextureRegionDrawable;
import com.nopalsoft.thetruecolor.Assets;
import com.badlogic.gdx.scenes.scene2d.ui.ImageButton;

public class Person extends Group implements Comparable<Person> {
    final float WIDTH = DialogRanking.WIDTH - 5;
    final float HEIGHT = 75;

    public enum AccountType {
        GOOGLE_PLAY, FACEBOOK, AMAZON
    }

    public Person.AccountType accountType;

    final public String id;
    public String name;
    public long score;

    Label labelName;
    Label labelScore;


    public Person(Person.AccountType accountType, String id, String name, long score) {
        setBounds(0, 0, WIDTH, HEIGHT);

        this.accountType = accountType;
        this.id = id;
        this.name = name;
        this.score = score;

        Image imageAccount = getImage(accountType);

        labelName = new Label(name, new LabelStyle(Assets.fontSmall, Color.BLACK));
        labelName.setFontScale(.7f);
        labelName.setPosition(140, 36);

        labelScore = new Label(formatScore(), new LabelStyle(Assets.fontSmall, Color.RED));
        labelScore.setPosition(140, 5);

        addActor(imageAccount);
        addActor(labelName);
        addActor(labelScore);

        // Separator
        Image imageHeader = new Image(Assets.header);
        imageHeader.setPosition(0, 0);
        imageHeader.setSize(WIDTH, 5);
        addActor(imageHeader);
    }

    private Image getImage(Person.AccountType accountType) {
       TextureRegionDrawable accountKey;
        switch (accountType) {
            case AMAZON:
                accountKey = Assets.buttonAmazon;
                break;
            case FACEBOOK:
                accountKey = Assets.buttonFacebook;
                break;
            default:
            case GOOGLE_PLAY:
                accountKey = Assets.buttonGoogle;
                break;
        }

        Image imageAccount = new Image(accountKey);
        imageAccount.setSize(30, 30);
        imageAccount.setPosition(10, HEIGHT / 2f - imageAccount.getHeight() / 2f);
        return imageAccount;
    }

    public void setPicture(TextureRegionDrawable drawable) {

        // I use an image button because it can have a background and an image.
        ImageButton imageButtonCharacter = new ImageButton(new ImageButton.ImageButtonStyle(drawable, null, null, Assets.photoFrame, null, null));
        imageButtonCharacter.setSize(60, 60);
        imageButtonCharacter.getImageCell().size(60);
        imageButtonCharacter.setPosition(60, HEIGHT / 2f - imageButtonCharacter.getHeight() / 2f);
        addActor(imageButtonCharacter);

    }


    public String formatScore() {
        String score = String.valueOf(this.score);
        int floatPos = score.contains(".") ? score.length() - score.indexOf(".") : 0;
        int nGroups = (score.length() - floatPos - 1 - (score.contains("-") ? 1 : 0)) / 3;
        for (int i = 0; i < nGroups; i++) {
            int commaPos = score.length() - i * 4 - 3 - floatPos;
            score = score.substring(0, commaPos) + "," + score.substring(commaPos);
        }
        return score;
    }

    @Override
    public int compareTo(Person otherPerson) {
        return Long.compare(otherPerson.score, score);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Person) {
            Person objPerson = (Person) obj;
            return id.equals(objPerson.id) && accountType == objPerson.accountType;

        } else
            return false;
    }

    public void updateData(String _name, long _score) {
        name = _name;
        score = _score;
        labelName.setText(name);
        labelScore.setText(formatScore());

    }
}

File: core/src/com/nopalsoft/thetruecolor/leaderboard/Person.kt
package com.nopalsoft.thetruecolor.leaderboard

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.scenes.scene2d.Group
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.ui.ImageButton
import com.badlogic.gdx.scenes.scene2d.ui.ImageButton.ImageButtonStyle
import com.badlogic.gdx.scenes.scene2d.ui.Label
import com.badlogic.gdx.scenes.scene2d.utils.TextureRegionDrawable
import com.nopalsoft.thetruecolor.Assets

class Person(val accountType: AccountType, val id: String, name: String, var score: Long) : Group(),
    Comparable<Person> {

    private val labelName = Label(name, Label.LabelStyle(Assets.fontSmall, Color.BLACK))
    private val labelScore = Label(formatScore(), Label.LabelStyle(Assets.fontSmall, Color.RED))


    init {

        setBounds(0f, 0f, WIDTH, HEIGHT)
        val imageAccount = getImage(accountType)
        labelName.setFontScale(.7f)
        labelName.setPosition(140f, 36f)
        labelScore.setPosition(140f, 5f)

        addActor(imageAccount)
        addActor(labelName)
        addActor(labelScore)


        // Separator
        val imageHeader = Image(Assets.header)
        imageHeader.setPosition(0f, 0f)
        imageHeader.setSize(WIDTH, 5f)
        addActor(imageHeader)
    }


    private fun getImage(accountType: AccountType): Image {
        val accountKey = when (accountType) {
            AccountType.AMAZON -> Assets.buttonAmazon
            AccountType.FACEBOOK -> Assets.buttonFacebook
            AccountType.GOOGLE_PLAY -> Assets.buttonGoogle
            else -> Assets.buttonGoogle
        }
        val imageAccount = Image(accountKey)
        imageAccount.setSize(30f, 30f)
        imageAccount.setPosition(10f, HEIGHT / 2f - imageAccount.height / 2f)
        return imageAccount
    }

    fun setPicture(drawable: TextureRegionDrawable) {

        // I use an image button because it can have a background and an image.
        val imageButtonCharacter =
            ImageButton(ImageButtonStyle(drawable, null, null, Assets.photoFrame, null, null))
        imageButtonCharacter.setSize(60f, 60f)
        imageButtonCharacter.imageCell.size(60f)
        imageButtonCharacter.setPosition(60f, HEIGHT / 2f - imageButtonCharacter.height / 2f)
        addActor(imageButtonCharacter)
    }

    fun formatScore(): String {
        var score = score.toString()
        val floatPos = if (score.contains(".")) score.length - score.indexOf(".") else 0
        val nGroups = (score.length - floatPos - 1 - (if (score.contains("-")) 1 else 0)) / 3
        for (i in 0 until nGroups) {
            val commaPos = score.length - i * 4 - 3 - floatPos
            score = score.substring(0, commaPos) + "," + score.substring(commaPos)
        }
        return score
    }


    override fun compareTo(other: Person) = other.score.compareTo(score)

    override fun equals(other: Any?): Boolean {
        return if (other is Person) {
            id == other.id && accountType == other.accountType
        } else false
    }

    fun updateData(_name: String, _score: Long) {
        name = _name
        score = _score
        labelName.setText(name)
        labelScore.setText(formatScore())
    }

    override fun hashCode(): Int {
        var result = accountType.hashCode()
        result = 31 * result + id.hashCode()
        result = 31 * result + score.hashCode()
        result = 31 * result + labelName.hashCode()
        result = 31 * result + labelScore.hashCode()
        return result
    }

    enum class AccountType {
        GOOGLE_PLAY, FACEBOOK, AMAZON
    }

    companion object {
        private const val WIDTH = DialogRanking.WIDTH - 5
        private const val HEIGHT = 75f
    }
}